Sources/Purchasing/StoreKit2/Observer Mode/StoreKit2ObserverModePurchaseDetector.swift

```diff
@@ -0,0 +1,111 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  StoreKit2ObserverModePurchaseDetector.swift
+//
+//  Created by Will Taylor on 5/1/24.
+
+import Foundation
+import StoreKit
+
+/// A delegate protocol for handling verified transactions in observer mode.
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+// swiftlint:disable type_name
+protocol StoreKit2ObserverModePurchaseDetectorDelegate: AnyObject, Sendable {
+
+    /// Handles a verified transaction with its corresponding JWS representation.
+    /// - Parameters:
+    ///   - verifiedTransaction: The verified transaction to be processed.
+    ///   - jwsRepresentation: The JSON Web Signature representation of the transaction.
+    func handleSK2ObserverModeTransaction(
+        verifiedTransaction: StoreKit.Transaction,
+        jwsRepresentation: String
+    ) async throws
+}
+
+/// Protocol describing an actor capable of detecting purchases from StoreKit 2.
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+protocol StoreKit2ObserverModePurchaseDetectorType: Actor {
+    func detectUnobservedTransactions(
+        delegate: StoreKit2ObserverModePurchaseDetectorDelegate
+    ) async
+}
+
+/// Actor responsibile for detecting purchases from StoreKit2 that should be processed by observer mode.
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+actor StoreKit2ObserverModePurchaseDetector: StoreKit2ObserverModePurchaseDetectorType {
+
+    private let deviceCache: DeviceCache
+    private let allTransactionsProvider: AllTransactionsProviderType
+
+    init(
+        deviceCache: DeviceCache,
+        allTransactionsProvider: AllTransactionsProviderType
+    ) {
+        self.deviceCache = deviceCache
+        self.allTransactionsProvider = allTransactionsProvider
+    }
+
+    /// Detects unobserved transactions and forwards them to the StoreKit2ObserverModeManagerDelegate
+    /// for processing.
+    func detectUnobservedTransactions(
+        delegate: StoreKit2ObserverModePurchaseDetectorDelegate
+    ) async {
+        guard let mostRecentTransaction = await allTransactionsProvider.getMostRecentVerifiedTransaction() else {
+            return
+        }
+
+        let jwsRepresentation = mostRecentTransaction.jwsRepresentation
```

Comment:
The old comment is now hidden because the code changed, but it's a good point that in practice we only need to detect a single change in the history to tell the backend and it can go and find any / all transactions that might have been missed. 

One thing that I do still wonder about, wouldn't we still have false positives if multiple transactions happen while the app is closed? 

I.e.:
1. make purchase, T1
1. come back after a few months. T2, T3 happened
1. re-open app, find T3, post, mark it synced
1. buy consumable, T4, sync, mark it synced. 
1. re-open app, we look at latest, T4, sync, mark it synced. T3 is now latest.
1. T3 revoked through family sharing
1. T2 is now latest, we post it, but it's been synced since step 3


Obviously this is a very edge-casey thing, but I imagine there's probably a lot of cases that I'm not thinking about off the top of my head. 

Maybe the safest approach is to consider all existing jwts at the time of posting the receipt as synced? 

In terms of how it differs to this approach, it'd mean: 
- instead of getting the most recent transaction, get all transactions and see if there are _any_ that we haven't marked synced (we're already fetching all from StoreKit and DeviceCache so it wouldn't be less efficient other than the in-memory comparison of two sets). 
- store the list of all "unsynced" transaction ids
- after getting a <5xx response from the backend, mark all unsynced transaction ids as synced

You can see an example of us doing a similar thing (minus the StoreKit part) that for subscriber attributes: 
https://github.com/RevenueCat/purchases-ios/blob/main/Sources/SubscriberAttributes/SubscriberAttributesManager.swift#L148

Reply:
I think this could also occur with a consumable that is finished in addition to a revoked family sharing transaction. I'll  get this implemented!

Reply:
@aboedo We're now marking all found transactions as synced to avoid this edge case, see https://github.com/RevenueCat/purchases-ios/pull/3853/commits/8a470de1a4183dec17c71632a63ec1666cc7b1ef 😎

Reply:
would it still make sense to look at all transactions from SK2 to find an unsynced one rather than latest in this case?